manta-trusted-setup/src/groth16/bn254/ppot.rs

```diff
@@ -0,0 +1,1102 @@
+// Copyright 2019-2022 Manta Network.
+// This file is part of manta-rs.
+//
+// manta-rs is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// manta-rs is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with manta-rs.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Utility functions for Perpetual Powers of Tau
+
+use crate::{
+    groth16::kzg::{
+        Accumulator, Configuration as KzgConfiguration, G1Marker, G2Marker, Proof as KzgProof, Size,
+    },
+    ratio::RatioProof,
+    util::{BlakeHasher, Deserializer, KZGBlakeHasher, Serializer},
+};
+use ark_bn254::{Bn254, Fr, G1Affine, G2Affine, Parameters};
+use ark_serialize::{CanonicalSerialize, Read, SerializationError, Write};
+use ark_std::{io, println};
+use blake2::Digest;
+use core::fmt;
+use manta_crypto::arkworks::{
+    ec::{
+        models::{bn::BnParameters, ModelParameters},
+        short_weierstrass_jacobian::GroupAffine,
+        AffineCurve, PairingEngine, SWModelParameters,
+    },
+    ff::{PrimeField, ToBytes, Zero},
+    pairing::Pairing,
+};
+use manta_util::{cfg_iter, into_array_unchecked, vec::Vec};
+use memmap::Mmap;
+
+#[cfg(feature = "rayon")]
+use manta_util::rayon::prelude::ParallelIterator;
+
+/// Configuration of the Perpetual Powers of Tau ceremony
+pub struct PpotCeremony;
```

Comment:
How much code is shared between this configuration and `MantaPaySetupCeremony`? If the manta pay system is just a subaccumulator we should make this explicit somehow and inherit as much as we can from the ceremony code.

Reply:
Their implementations of `Pairing` and `groth16::kzg::Configuration` are identical. Is a macro the best way to do this?

Reply:
If the only difference is the size then you can do something like this:

```rust
pub struct PerpetualPowersOfTauCeremony<const POWERS: usize>;
```

and then give the implementations for this type instead.

Reply:
That works well for implementations of `Pairing` and `groth16::kzg::Configuration` but not for `Serializer` and `Deserializer` because we want to use arkworks canonical serialization for the `MantaPaySetupCeremony` but bellman-style serialization for the original ppot ceremony.  Any suggestion there?

Reply:
When in doubt, add more paramters!

```rust
pub struct PerpetualPowersOfTauCeremony<S, const POWERS: usize>;
```

where `S` implements the appropriate `Serializer` and `Deserializer`.